### PR TITLE
[BEAM-1427] BigQueryIO should comply with PTransform style guide

### PR DESCRIPTION
**Pull request by jkff**
Main (backward incompatible) changes:
- Classes BigQueryIO.Read.Bound and Write.Bound are now called simply BigQueryIO.Read and Write.
- Previously all parameters could be specified both via static and non-static methods. Now, BigQueryIO.Read.from() and BigQueryIO.Write.to() are now the only static factory methods (and they are available _only_ as static methods), and all other methods are non-static.

A bunch of code is saved by using AutoValue.

R: @tgroh  
CC: @dhalperi @sammcveety (Sam FYI - I think there shouldn't be too many conflicts with your PR)

### CODE REVIEW

**Comment by jkff** [comment]
Per offline discussion:
R: @reuvenlax 
and subsequently R: @tgroh for committer once Reuven LGTM's.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 1]
```diff
@@ -602,63 +588,65 @@ public void validate(PBegin input) {
...
-              String.format(QUERY_VALIDATION_FAILURE_ERROR, query.get()), e);
+              String.format(QUERY_VALIDATION_FAILURE_ERROR, getQuery().get()), e);
         }
       }
     }
 
     @Override
     public PCollection<TableRow> expand(PBegin input) {
-      stepUuid = randomUUIDString();
+      String stepUuid = randomUUIDString();
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Sure this is not used anywhere else?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 2]
```diff
@@ -1724,23 +1637,25 @@ public void validate(PCollection<TableRow> input) {
...
       // Require a schema if creating one or more tables.
       checkArgument(
-          createDisposition != CreateDisposition.CREATE_IF_NEEDED || jsonSchema != null,
+          getCreateDisposition() != CreateDisposition.CREATE_IF_NEEDED || getJsonSchema() != null,
           "CreateDisposition is CREATE_IF_NEEDED, however no schema was provided.");
 
       // The user specified a table.
-      if (jsonTableRef != null && validate) {
+      BigQueryServices bqServices =
+          MoreObjects.firstNonNull(getBigQueryServices(), new BigQueryServicesImpl());
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: if getBigQueryService() returns non null, we still create a throwaway object here, correct?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 3]
```diff
@@ -2789,11 +2762,16 @@ public PDone expand(PCollection<TableRow> input) {
...
-              createDisposition,
-              tableDescription,
-              bqServices)));
+          .apply(
+              ParDo.of(
+                  new StreamingWriteFn(
+                      write.getJsonSchema() == null
+                          ? null
+                          : NestedValueProvider.of(
+                              write.getJsonSchema(), new JsonSchemaToTableSchema()),
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Pull this conditional assignment into a member variable  - kinda hard to read here.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 1 above)
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Correct, it's only used in this method.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 2 above)
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: It's already fixed in one of the following commits.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 3 above)
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done.

**Comment by reuvenlax** [comment]
Looks good to me.

**Comment by jkff** [comment]
retest this please

**Comment by jkff** [comment]
@tgroh - Reuven LGTMd, can you take a skim and commit if it's fine?

**Pull request closed without merging**